src/nn/visit_tensors.rs

```diff
@@ -0,0 +1,178 @@
+use crate::shapes::Dtype;
+
+use crate::prelude::*;
+use std::{
+    fmt::Debug,
+    string::{String, ToString},
+};
+
+pub struct ModuleGroup<'a, const N: usize, const M: usize, T> {
+    pub refs: [&'a T; N],
+    pub refs_mut: [&'a mut T; M],
+    pub name: Option<String>,
+}
+
+// TODO? : Prevent heap allocation using unsafe or ArrayVec
+fn map_mut<'a, A: 'a + Debug, B: 'a + Debug, const N: usize, F: FnMut(&'a mut A) -> &'a mut B>(
+    arr: &'a mut [A; N],
+    func: F,
+) -> [&'a mut B; N] {
+    let vec: std::vec::Vec<_> = arr.iter_mut().map(func).collect();
+
+    vec.try_into().unwrap()
+}
+
+impl<'a, const N: usize, const M: usize, T: Debug> ModuleGroup<'a, N, M, T> {
+    pub fn new(refs: [&'a T; N], refs_mut: [&'a mut T; M], name: Option<String>) -> Self {
+        Self {
+            refs,
+            refs_mut,
+            name,
+        }
+    }
+
+    pub fn map<T2: Debug, F1: FnMut(&T) -> &T2, F2: FnMut(&mut T) -> &mut T2>(
+        &mut self,
+        func1: F1,
+        mut func2: F2,
+        name: &str,
+    ) -> ModuleGroup<N, M, T2> {
+        ModuleGroup {
+            refs: self.refs.map(func1),
+            refs_mut: map_mut(&mut self.refs_mut, |x| func2(*x)),
+            name: self
+                .name
+                .as_ref()
+                .map(|prefix| std::format!("{prefix}{name}")),
+        }
+    }
+
+    pub fn visit<E: Dtype, D: DeviceStorage, F: TensorVisitor<N, M, E, D>>(
+        self,
+        func: &mut F,
+    ) -> Result<(), F::Err>
+    where
+        T: VisitTensorGroups<N, M, E, D>,
+    {
+        VisitTensorGroups::visit_groups(self, func)
+    }
+}
+
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum TensorVisitorOption {
+    DisableGradientUpdate,
+}
+
+pub trait TensorVisitor<const N: usize, const M: usize, E: Dtype, D: DeviceStorage> {
+    type Err: std::fmt::Display + Debug;
+
+    fn call<S: Shape>(
+        &mut self,
+        tensors: ModuleGroup<N, M, Tensor<S, E, D>>,
+        option: &[TensorVisitorOption],
+    ) -> Result<(), Self::Err>;
+}
+
+pub trait VisitTensorGroups<const N: usize, const M: usize, E: Dtype, D: DeviceStorage>:
+    Sized
+{
+    fn visit_groups<F: TensorVisitor<N, M, E, D>>(
+        self_refs: ModuleGroup<N, M, Self>,
+        func: &mut F,
+    ) -> Result<(), F::Err>;
+}
+
+impl<const N: usize, const M: usize, S: Shape, E: Dtype, D: DeviceStorage>
+    VisitTensorGroups<N, M, E, D> for Tensor<S, E, D>
+{
+    fn visit_groups<F: TensorVisitor<N, M, E, D>>(
+        self_refs: ModuleGroup<N, M, Self>,
+        func: &mut F,
+    ) -> Result<(), F::Err> {
+        func.call(self_refs, &[])
+    }
+}
+
+pub trait VisitTensors<E: Dtype, D: DeviceStorage>: VisitTensorGroups<1, 0, E, D> + Debug {
```

Comment:
Are `1, 0` and `0, 1` for VisitTensorsRef and VisitoTensorsMut the only two cases we would ever support?

I think we might be able to get rid of the VisitTensorGroups trait if we can separately impl some trait (not sure which one) for &T and &mut T

Reply:
We will eventually support `1, 1` so we can add an multiply modules for EMA.